core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,253 @@
+        // writers if it is scaled.
+        for (int partition = 0; partition < partitionAssignments.length; partition++) {
+            IntArrayList positionsList = partitionAssignments[partition];
+            int partitionSize = positionsList.size();
+            if (partitionSize == 0) {
+                continue;
+            }
+
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            int nextPartitionWriter = getNextPartitionWriter(partition, partitionPhysicalWrittenBytes);
+
+            for (int i = 0; i < partitionSize; i++) {
+                writerAssignments[nextPartitionWriter].add(positions[i]);
+            }
+        }
+
+        // build a page for each writer
+        for (int bucket = 0; bucket < writerAssignments.length; bucket++) {
+            IntArrayList positionsList = writerAssignments[bucket];
+            int bucketSize = positionsList.size();
+            if (bucketSize == 0) {
+                continue;
+            }
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            if (bucketSize == page.getPositionCount()) {
+                // entire page will be sent to this writer.
+                return buffers.get(bucket);
+            }
+            Page pageSplit = page.copyPositions(positions, 0, bucketSize);
+            sendPageToPartition(buffers.get(bucket), pageSplit);
+        }
+
+        // No single writer receives the entire input page
+        return null;
+    }
+
+    private int getNextPartitionWriter(int partition, long[] partitionPhysicalWrittenBytes)
+    {
+        PartitionScalingInfo scalingInfo = partitionScalingInfos[partition];
+        int maxWriterCount = scalingInfo.getMaxWriterCount();
+
+        // Scale up writers for a partition when current buffer memory utilization is more than 50% of the
+        // maximum and physical written bytes is greater than writerMinSize.
+        // This also mean that we won't scale local writers if the writing speed can cope up
+        // with incoming data. In another word, buffer utilization is below 50%.
+        if (maxWriterCount <= (buffers.size() - initialWriterCount) && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
+            long physicalWrittenBytes = partition < partitionPhysicalWrittenBytes.length ? partitionPhysicalWrittenBytes[partition] : 0;
```

Comment:
How this will detect that we have heavy used partition? same with `physicalWrittenBytes >= writerMinSize * maxWriterCount` after some time (and size) we will just start writers if we have memory buffers filled.